cocos/base/CCEventDispatcher.cpp

```diff
@@ -1313,20 +1390,40 @@ void EventDispatcher::sortEventListeners(const EventListener::ListenerID& listen
 
 void EventDispatcher::sortEventListenersOfSceneGraphPriority(const EventListener::ListenerID& listenerID, Node* rootNode)
 {
+
     auto listeners = getListeners(listenerID);
-    
     if (listeners == nullptr)
         return;
     auto sceneGraphListeners = listeners->getSceneGraphPriorityListeners();
-    
     if (sceneGraphListeners == nullptr)
         return;
 
-    // Reset priority index
-    _nodePriorityIndex = 0;
-    _nodePriorityMap.clear();
 
-    visitTarget(rootNode, true);
+	// Reset priority index
+	_nodePriorityIndex = 0;
+	_nodePriorityMap.clear();
+
+	// reverse visit, from leaf to root, reduce the time of tree's travel
+	// use node userdata to mark, speed up hash find
+	for (const auto l : *sceneGraphListeners)
+	{
+		Node* node = l->getAssociatedNode();
+		if (node)
+			node->setUserData((void*)1);
```

Comment:
You can not setUserData.
It can be used by user.

Reply:
yes, I known its not a right way which use UserData to storage some visit information for engine. and its simple to add an "SystemData" in Node for optimize